Memoize unary, binary and func expressions.

petermattis opened the pull request:
Fixed a bug where "SELECT col+1 FROM table" would return an error if col
was NULL instead of evaluating to NULL.

Reorganized the scanNode initialization code so that it is mostly done
in scanNode instead of planner.Select.

petermattis committed:
Reorganized the scanNode initialization code.

Initialize is now mostly done in scanNode instead of planner.Select.

tamird commented on sql/scan.go:
```diff
@@ -102,24 +112,68 @@ func (n *scanNode) Err() error {
-		n.render[i], n.err = n.extractQVals(n.render[i])
-		if n.err != nil {
-			return false
+	case 1:
```
this looks like a lot of code movement that is unrelated to the "real" change in this PR. can it at least get its own commit?

petermattis replied:
I broke out the code movement into a separate commit. PTAL.

tamird commented on sql/parser/eval.go:
```diff
@@ -627,12 +629,25 @@ func evalBinaryExpr(expr *BinaryExpr) (Datum, error) {
+		expr.ltype = ltype
+		expr.rtype = rtype
+	}
+	if expr.fn == nil {
```
this is set immediately above, how can it even be nil here?

tamird replied:
oh it's a hash lookup. disregard.

tamird commented on sql/parser/eval.go:
```diff
@@ -667,23 +686,41 @@ func evalFuncExpr(expr *FuncExpr) (Datum, error) {
-		if candidate.match(types) {
-			impl = candidate.fn
-			break
+	needMatch := true
```
i think this would work just the same without needMatch

petermattis replied:
Yeah, it was just a small optimization to avoid performing the match a second time when not necessary. Removed for now.

petermattis committed:
Memoize unary, binary and func expressions.

Fixed a bug where "SELECT col+1 FROM table" would return an error if col
was NULL instead of evaluating to NULL.

tamird commented:
LGTM
